### PR TITLE
NOT READY TO MERGE: Media Store Integration

### PR DESCRIPTION
**Pull request by oguzkocer**
This PR integrates `MediaStore` from [FluxC library](https://github.com/wordpress-mobile/WordPress-FluxC-Android). The integration is not fully done, but I wanted to open a PR for us to discuss some integration details asynchronously.

I think vast majority of the getters have been updated to use `MediaStore` if not all of them. If you find any I missed, please do let me know. I was able to use `Dispatcher` for some features, but most of that is missing. I want to keep an up-to-date todo list here, so the rest of integration is handled in a more structured way. I've also added some todos in the code, because the library didn't fully support what I wanted to do and I didn't want to forget about them.

My main struggle right now is updating delete & upload flows, I could really use some help there @tonyr59h whenever you're available.

### CODE REVIEW

**Comment by tonyr59h** [comment]
Taking a look at the upload flow atm but might be touching some of `MediaDeleteService` as well. Going to convert our current `Service`'s to `IntentService`'s as suggested in [the docs](https://developer.android.com/topic/performance/memory.html#Services).

**Comment by maxme** [comment]
@tonyr59h what is needed before this one is ready for review? I think this is the only big thing remaining before we do a full check, we should all focus on this.

**Comment by oguzkocer** [comment]
I was actually going to comment something similar today, @maxme beat me to it. What I was going to suggest was, get this PR building and at least partially working and create new issues/checklist for what's missing. This way we should all be able to help out and work this out asap.

**Comment by maxme** [comment]
I made minor changes in this PR and I did a first test pass, here are the first thing we must fix. It would be a good idea to start by FluxC stuff.

# Media Library

- [ ] Delete media
- [ ] Implement media upload (MediaBrowserActivity.queueFileForUpload and Fix MediaUploadService)
- [ ] FETCH_ALL_MEDIA sometimes never reply
- [ ] Delete some medias remotely, then refresh the Media Library. Deleted medias are still in the list.
- [ ] BUG: got a "media must define a valid MIME type" error after trying to upload a valid image (from Photos).
- [ ] Implement paging (currently limited to first 20)

# Post Editor

- [ ] Implement media upload

# FluxC Media

- [ ] `MediaStore.performUploadMedia` should return a OnMediaUploaded event when an error occurs before actual upload - it currently returns a OnMediaChanged event with an error ("Media error occurred: MALFORMED_MEDIA_ARG" - invalid mimetype for instance)
- [ ] Like the CommentStore / PostStore, we have to drop all medias related to a Site when we received a FETCHED_ALL_MEDIA to handle deletion on the server.

**Comment by tonyr59h** [comment]
> we have to drop all medias related to a Site when we received a FETCHED_ALL_MEDIA to handle deletion coming from the server

There might be situations that remove media inconveniently, for instance any media queued in the browser would have to be re-queued if a refresh occurs.

**Comment by maxme** [comment]
> There might be situations that remove media inconveniently, for instance any media queued in the browser would have to be re-queued if a refresh occurs.

The queue should probably be kept on the app side and not stored in the media store, so deleting data from the media store won't affect it.

I'm not sure how we can handle deletion from the server if we don't do that (we can't check every id in the table, and there is no "history" of events on the server).

**Comment by tonyr59h** [comment]
> It stays like that even after multiple relaunches of the app, not sure what's going on.

I'm pretty sure the progress indicator is optimistic to a point and shows some 'fake progress' at the start of an upload. cc @aforcier

> The queue should probably be kept on the app side and not stored in the media store, so deleting data from the media store won't affect it.

Gotcha, added the changes to `issue/missing-integration-functionality` in FluxC.

**Comment by aforcier** [comment]
>> It stays like that even after multiple relaunches of the app, not sure what's going on.

> I'm pretty sure the progress indicator is optimistic to a point and shows some 'fake progress' at the start of an upload. cc @aforcier

That's right, optimistic progress will continue till ~15% and freeze there, regardless of actual upload progress.

**Comment by maxme** [comment]
> Implement paging (currently limited to first 20) - not just for the first 2 pages. - done in 7245cd2

@tonyr59h I don't think this is working, when I scroll down in the Media library, it's not fetching new medias, but when I pull to refresh several times, it seems to fetch and display new pages.

**Comment by maxme** [comment]
> Implement media upload (MediaBrowserActivity.queueFileForUpload and Fix MediaUploadService) - @tonyr59h

@tonyr59h This is almost done, I'll let you finish if you want and have a look at sorting issues.

**Comment by tonyr59h** [comment]
The upload/delete media services are up and running. You can test them in the media browser, though there is still the issue of the adapter not refreshing after the upload completes. Coming soon to post editor.

**Comment by tonyr59h** [comment]
The issue with FETCH_ALL_MEDIA has been resolved via [FluxC changes](https://github.com/wordpress-mobile/WordPress-FluxC-Android/commit/93d16917356ada6d28810b7f875fb421397e61ec). We may still want to paginate better but the change seemed appropriate given the action name.

**Comment by maxme** [comment]
> The issue with FETCH_ALL_MEDIA has been resolved via FluxC changes. We may still want to paginate better but the change seemed appropriate given the action name.

That looks good for wpcom sites, but it seems broken for self hosted sites: I get into a loop of `wp.getMediaLibrary` and the results seems ignored the medial library remains empty.

**Comment by maxme** [comment]
@tonyr59h The media library looks OK, seems feature complete compared to develop. What's remaining in the editor?

**Comment by tonyr59h** [comment]
> What's remaining in the editor?

Upload and retrying are working in general. Error paths, self-hosted, and videos still need to be worked through and more changes will need to be integrated in FluxC for some of it.

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 1]
```diff
@@ -1936,10 +1936,11 @@ private MediaFile queueFileForUpload(String path, ArrayList<Long> mediaIdOut, St
         media.setFileName(filename);
         media.setFilePath(path);
         media.setSiteId(mSite.getSiteId());
+        media.setMediaId(System.currentTimeMillis());
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: I don't think we should set the id locally

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 2]
```diff
@@ -1936,10 +1936,11 @@ private MediaFile queueFileForUpload(String path, ArrayList<Long> mediaIdOut, St
...
         media.setFileExtension(fileExtension);
         media.setMimeType(mimeType);
         media.setUploadState(MediaUploadState.QUEUED.toString());
-        media.setUploadDate(DateTimeUtils.iso8601UTCFromTimestamp(System.currentTimeMillis() / 1000));
+        media.setUploadDate(DateTimeUtils.iso8601UTCFromTimestamp(media.getMediaId() / 1000));
```
Comment (maxme) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java]: And we should set this back, this is weird.

**Comment by tonyr59h** [comment]
### Wrapping it up
- [ ] [FluxC PR](https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/274) merged
- [ ] Video capture flow in both browser and editor needs to be verified
- [ ] Persisted media service queue's (WIP)
- [ ] Retry error handling (media not in store, store media doesn't match remote, etc...)
- [ ] Paging in media browser
- [ ] Self-hosted fixes in both editor and browser

**Comment by maxme** [comment]
We're going to merge this PR even if there are known issues, it's becoming a monster and very hard to work on this with more than one person. 

The idea is to open a ticket for each point above and for other related issues we encounter with media in the integration branch. We can work on them without stepping on each other toes.

**Comment by maxme** [comment]
Created the following tickets:

https://github.com/wordpress-mobile/WordPress-Android/issues/5176
https://github.com/wordpress-mobile/WordPress-Android/issues/5177
https://github.com/wordpress-mobile/WordPress-Android/issues/5178
https://github.com/wordpress-mobile/WordPress-Android/issues/5179

**Comment by maxme** [comment]
🚢 🚢 🚢 See you in other tickets Media!

**Pull request merged**